File: src/backend/b_openvino.cpp
#include "backend/b_openvino.hpp"

#include <cassert>
#include <cstdint>
#include <iostream>
#include <memory>
#include <string>
#include <vector>

#include "backend/interface.hpp"
#include "opencv2/core/mat.hpp"
#include "opencv2/imgproc.hpp"
#include "utils/extend_print.hpp"

OpenVINOBackend::OpenVINOBackend(const det_input_t& input_params) : BaseDetectionBackend(input_params) {
    auto devices = this->GetDevice();
    std::string devices_str;
    for (auto&& device : devices) {
        devices_str += device + ", ";
    }
    OPENVINO_ASSERT(std::find(devices.begin(), devices.end(), this->input_params.device) != devices.end(),
                    "Device name is not in available devices: " + devices_str);
}

OpenVINOBackend::~OpenVINOBackend() {}

void OpenVINOBackend::LoadModel(const std::string& model_path) {
    ov::Core core;
    std::shared_ptr<ov::Model> model = core.read_model(model_path);

    OPENVINO_ASSERT(model->inputs().size() == 1, "Sample supports models with 1 input only");
    OPENVINO_ASSERT(model->outputs().size() == 1, "Sample supports models with 1 output only");
    // -------- Step 3. Set up input

    ov::element::Type input_type = ov::element::u8;
    const ov::Layout tensor_layout{"NHWC"};

    // -------- Step 4. Configure preprocessing --------

    ov::preprocess::PrePostProcessor ppp(model);

    // 1) Set input tensor information:
    // - input() provides information about a single model input
    // - reuse precision and shape from already available `input_tensor`
    // - layout of data is 'NHWC'
    ppp.input().tensor().set_element_type(input_type).set_layout(tensor_layout);
    // 2) Adding explicit preprocessing steps:
    // - convert layout to 'NCHW' (from 'NHWC' specified above at tensor layout)
    // - apply linear resize from tensor spatial dims to model spatial dims
    ppp.input().preprocess().resize(ov::preprocess::ResizeAlgorithm::RESIZE_LINEAR);
    // 4) Here we suppose model has 'NCHW' layout for input
    ppp.input().model().set_layout("NCHW");
    // 5) Set output tensor information:
    // - precision of tensor is supposed to be 'f32'
    ppp.output().tensor().set_element_type(ov::element::f32);

    // 6) Apply preprocessing modifying the original 'model'
    model = ppp.build();

    // -------- Step 5. Loading a model to the device --------
    this->compiled_model = core.compile_model(model, this->input_params.device);
    if (this->compiled_model.input().get_shape().size() == 4) {
        auto input_shape = this->compiled_model.input().get_shape();
        this->input_params.batch_size = input_shape[0];
        this->input_params.height = input_shape[1];
        this->input_params.width = input_shape[2];
        this->input_params.channel = input_shape[3];
    } else
        throw std::runtime_error("Model input shape is not BCHW");
    if (this->compiled_model.output().get_shape().size() != 3)
        throw std::runtime_error("Model output shape is not BNS");
    auto output_shape = this->compiled_model.output().get_shape();
    // this->batch_predictions = det_output_t(output_shape[0], std::vector<detetcion_output_t>(output_shape[1]));
}

det_output_t OpenVINOBackend::Infer(const cv::Mat& _img) {
    cv::Mat img;
    cv::resize(_img,
               img,
               cv::Size(this->compiled_model.input().get_shape()[1], this->compiled_model.input().get_shape()[2]));

    auto infer_request = compiled_model.create_infer_request();
    infer_request.set_input_tensor(
        ov::Tensor(compiled_model.input().get_element_type(), compiled_model.input().get_shape(), img.data));
    infer_request.infer();

    const ov::Tensor& output_tensor = infer_request.get_output_tensor();
    auto output_shape = output_tensor.get_shape();

    float* detections = output_tensor.data<float>();
    det_output_t batch_predictions;
    for (int b = 0; b < output_shape[0]; b++) {
        std::vector<detetcion_output_t> predictions;
        std::vector<cv::Rect> boxes;
        for (int n = 0; n < output_shape[1]; n++) {
            float* detection = &detections[(b * output_shape[1] + n) * output_shape[2]];

            if (detection[4] < this->input_params.threshold.conf)
                continue;

            int max_class_id = 5;
            for (int i = 5; i < output_shape[2]; i++) {
                if (detection[i] > detection[max_class_id])
                    max_class_id = i;
            }
            if (detection[max_class_id] < this->input_params.threshold.score)
                continue;

            auto box = cv::Rect(cv::Point(detection[0], detection[1]), cv::Point(detection[2], detection[3]));
            detetcion_output_t prediction{
                box,
                detection[4],
                max_class_id - 5,
                detection[max_class_id],
            };
            boxes.push_back(box);

            predictions.emplace_back(prediction);
        }

        batch_predictions.emplace_back(predictions);
    }

    return batch_predictions;  // TODO: enable RVO
}

std::vector<std::string> OpenVINOBackend::GetDevice() const {
    ov::Core core;
    auto availableDevices = core.get_available_devices();
#ifdef DEBUG
    {
        for (auto&& device : availableDevices) {
            std::cout << device << std::endl;
        }
    }
#endif
    return availableDevices;
}

File: dlinfer/detector/b_openvino.py
import time
from .interface import IDetectoBackends
import numpy as np
import openvino as ov
from openvino.preprocess import PrePostProcessor
from openvino.runtime import (
    __version__,
    AsyncInferQueue,
    Core,
    CompiledModel,
    InferRequest,
    Layout,
    Type,
    Model,
)


class OpenVINODetectorBackend(IDetectoBackends):
    NAME = "OpenVINO"
    SUPPORTED_VERISONS = ["2023.0.1", "2024.1.0"]
    SUPPORTED_DEVICES = ["CPU", "GPU", "MYRIAD", "HDDL", "HETERO"]

    def __init__(self, device="AUTO") -> None:
        ov_version: str = __version__
        super().__init__(ov_version)
        self.core: Core = Core()
        self.device = device
        self.model: CompiledModel = None
        self.infer_request = None
        self.async_mode = False
        self.curr_request = None
        self.next_request = None
        # TODO： [Multi-device execution](https://docs.openvino.ai/2024/openvino-workflow/running-inference/inference-devices-and-modes/multi-device.html)

    def load_model(self, model_path: str, verbose: bool = False) -> None:
        # fmt: off
        model: Model = self.core.read_model(model_path)
        assert len(model.inputs) == 1, "Sample supports only single input topologies"
        assert len(model.outputs) == 1, "Sample supports only single output topologies"
        
        # -- Apply preprocessing
        # https://docs.openvino.ai/2024/openvino-workflow/running-inference/optimize-inference/optimize-preprocessing.html
        ppp = PrePostProcessor(model)

        # 1) Set input tensor information:
        # - input() provides information about a single model input
        # - precision of tensor is supposed to be 'u8'
        # - layout of data is 'NHWC'
        # ppp.input().tensor().set_element_type(Type.u8).set_layout(
        #     Layout("NHWC")
        # )  # noqa: N400

        # 2) Here we suppose model has 'NCHW' layout for input
        ppp.input().model().set_layout(Layout("NCHW"))

        # 3) Set output tensor information:
        # - precision of tensor is supposed to be 'f32'
        ppp.output().tensor().set_element_type(Type.f32)

        # 4) Apply preprocessing modifing the original 'model'
        model = ppp.build()

        # -- Loading model to the device
        compiled_model = self.core.compile_model(model, self.device)
        self.model = compiled_model
        self.infer_request = compiled_model.create_infer_request()
        self.input_layer_ir = model.input(0)
        # fmt: on

    def enable_async_mode_infer(self, completion_callback):
        infer_queue = AsyncInferQueue(
            self.model, jobs=2
        )  # API: https://docs.openvino.ai/2024/api/ie_python_api/_autosummary/openvino.runtime.AsyncInferQueue.html
        infer_queue.set_callback(completion_callback)
        self.infer_queue = infer_queue
        self.async_mode = True

    def infer(self, input: np.ndarray) -> np.ndarray:
        """
        # 执行模型推理，对 ONNX 模型进行推理的封装
        ## Args:
        - `input`: `np.ndarray`, `(1, C, H, W)`, `np.float32`, range: `[0, 1]`

        ## Returns:
        - `output`: `np.ndarray`, `(201, 18, 4)`, `np.float32`, range: `[0, 1]`
        """
        # `input name`  is compatible with ONNX model `binding__input_names`
        # `output name` is compatible with ONNX model `binding_output_names`
        # outputs: List[np.ndarray] = self.ort_session.run(["output"], {"input": input})
        # output = outputs[0]
        # return output
        input_layer_ir = self.input_layer_ir
        infer_request = self.infer_request
        infer_request.set_tensor(input_layer_ir, ov.Tensor(input))
        infer_request.infer()
        preds = infer_request.get_output_tensor(0).data
        return preds

    def infer_async(self, input: np.ndarray, **kwargs) -> np.ndarray:
        """
        https://docs.openvino.ai/2024/openvino-workflow/running-inference/integrate-openvino-with-your-application/inference-request.html#asynchronous-mode
        """
        infer_queue = self.infer_queue
        input_layer_ir = self.input_layer_ir
        infer_queue.start_async(
            {input_layer_ir.any_name: input},
            (kwargs),
        )

    @staticmethod
    def query_device():
        """Query available devices for OpenVINO backend."""
        return Core().available_devices


# input_layer_ir = model.input(0)
# N, C, H, W = input_layer_ir.shape
# shape = (H, W)
